### PR TITLE
Support user-defined s3 endpoint

### PR DESCRIPTION
**Pull request by luomiao**
This PR supports user-defined S3 endpoint.
By default, users should use AWS S3 for the state storage.
However, users may want to save the cluster state storage inside private cloud, through their own S3 services, such as [Minio](https://www.minio.io/).

With this PR, by setting environment variable "S3_ENDPOINT", kops will use the s3 service running on this specific s3 endpoint, instead of using AWS S3 by default. Accordingly, users should also update AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY to the corresponding credentials for this specific s3 service. It's the user's responsibility to setup their own s3 services.

*Update: Need to discuss whether we want AWS_ACCESS_KEY_ID or S3_ACCESS_KEY_ID. The reason for the latter is because another S3 server https://github.com/scality/S3 is using SCALITY_ACCESS_KEY_ID.

For example, one should start a Minio service on certain node by:
```
docker run -d \
    --name minio \
    -p 9000:9000 \
    -v /root/minio-data:/export \
    minio/minio server /export
```

Then create bucket:
`./mc config host add myminio http://127.0.0.1:9000 $accessKey $secretAccessKey
/mc mb myminio/[bucketname]`

When using this Minio service as the state store, one should just set:
`export S3_ENDPOINT=http:/[IP address of the Minio node]:9000
`

We assume that users who use their own user-defined S3 service, will not run the cluster with AWS EC2 service neither. Thus the credential information can be overwritten by the user-defined s3 service.

### CODE REVIEW

File: pkg/model/bootstrapscript.go [hunk 1]
```diff
@@ -61,31 +59,17 @@ func (b *BootstrapScript) ResourceNodeUp(ig *kops.InstanceGroup) (*fi.ResourceHo
...
+		// Pass in extra environment variables for user-defined S3 service
+		"S3Env": func() string {
+			if os.Getenv("S3_ENDPOINT") != "" {
+				return "export S3_ENDPOINT=" + os.Getenv("S3_ENDPOINT") + "\n" + "export AWS_REGION=" + os.Getenv("AWS_REGION") + "\n" + "export AWS_ACCESS_KEY_ID=" + os.Getenv("AWS_ACCESS_KEY_ID") + "\n" + "export AWS_SECRET_ACCESS_KEY=" + os.Getenv("AWS_SECRET_ACCESS_KEY")
 			}
 			return ""
 		},
 	}
 
 	nodeupTemplate := resources.AWSNodeUpTemplate
```
Comment (prashima) [inline pkg/model/bootstrapscript.go]: Nit: nodeupTemplate is not really needed now, resources.AWSNodeUpTemplate can be passed directly.

File: pkg/model/bootstrapscript.go [hunk 2]
```diff
@@ -61,31 +59,17 @@ func (b *BootstrapScript) ResourceNodeUp(ig *kops.InstanceGroup) (*fi.ResourceHo
...
-			return ""
-		},
-		"Env3": func() string {
-			if b.AddAwsEnvironmentVariables && os.Getenv("AWS_SECRET_ACCESS_KEY") != "" {
-				return "export AWS_SECRET_ACCESS_KEY=" + os.Getenv("AWS_SECRET_ACCESS_KEY")
+
+		// Pass in extra environment variables for user-defined S3 service
+		"S3Env": func() string {
+			if os.Getenv("S3_ENDPOINT") != "" {
+				return "export S3_ENDPOINT=" + os.Getenv("S3_ENDPOINT") + "\n" + "export AWS_REGION=" + os.Getenv("AWS_REGION") + "\n" + "export AWS_ACCESS_KEY_ID=" + os.Getenv("AWS_ACCESS_KEY_ID") + "\n" + "export AWS_SECRET_ACCESS_KEY=" + os.Getenv("AWS_SECRET_ACCESS_KEY")
```
Comment (prashima) [inline pkg/model/bootstrapscript.go]: Too many strings getting concatenated here. I would recommend using bytes.Buffer.

**Review by prashima — COMMENTED** [review]
Can you please update end-to-end testing, done for this change?

**Comment by msterin** [comment]
Should #31  be closed then ?

File: pkg/model/bootstrapscript.go (hunk 2 above)
Comment (msterin) [reply pkg/model/bootstrapscript.go]: or https://golang.org/pkg/fmt/#Sprintf

File: util/pkg/vfs/s3context.go [hunk 3]
```diff
@@ -61,6 +64,50 @@ func (s *S3Context) getClient(region string) (*s3.S3, error) {
...
+		session := session.New()
+		s3Client = s3.New(session, s3Config)
+		s.clients[region] = s3Client
+	}
+
+	return s3Client, nil
+}
+
+func (s *S3Context) setClientFunc() {
+	Endpoint := os.Getenv("S3_ENDPOINT")
```
Comment (msterin) [inline util/pkg/vfs/s3context.go]: since S3_ENDPOINT is used in more that 1 place would be nice to have it as a const

**Review by msterin — COMMENTED** [review]
LGTM  to my still untrained eye. 
A few nits are below. The only question - how it is tested and is there a way to add a CI test ?

File: pkg/model/bootstrapscript.go (hunk 2 above)
Comment (luomiao) [reply pkg/model/bootstrapscript.go]: done

File: pkg/model/bootstrapscript.go (hunk 1 above)
Comment (luomiao) [reply pkg/model/bootstrapscript.go]: done

File: util/pkg/vfs/s3context.go (hunk 3 above)
Comment (luomiao) [reply util/pkg/vfs/s3context.go]: It's only used twice. And considering the AWS_XXX envs are also getting called, maybe to be consistent we should just keep it like this?

Comment (msterin) [reply util/pkg/vfs/s3context.go]: we do not have to be consistent with something which is not done right :-)

But this is minor, OK either way.

**Pull request merged**